packages/syft/src/syft/service/network/network_service.py

```diff
@@ -237,19 +237,39 @@ def add_peer(
                 message="verify_key does not match the remote node's verify_key for add_peer"
             )
 
+        # check if the peer already is a node peer
+        existed_peer = self.stash.get_by_uid(context.node.verify_key, peer.id)
+        if existed_peer.is_err():
+            return SyftError(
+                message=f"Failed to query peer from stash: {existed_peer.err()}"
+            )
+        if isinstance(existed_peer.ok(), NodePeer):
+            return SyftError(
+                message=f"The peer '{peer.name}' is already associated with '{context.node.name}'"
+            )
+
+        # check if the peer already submitted an association request
+        association_requests: list[Request] = self._get_association_requests_by_peer_id(
+            context=context, peer_id=peer.id
+        )
+        if (
+            association_requests
+            and association_requests[-1].status == RequestStatus.PENDING
+        ):
+            return SyftError(
+                message="There is already a pending association request for this peer"
+            )
+        # only create and submit a new request if there is no requests yet
+        # or all previous requests have been rejected
```

Comment:
@kiendang please review this

Reply:
This is the fix for the association requests that we discussed earlier
- Check the last association request by timestamp (`updated_at`) and assign `PEER_ASSOCIATION_PENDING` only when the last request is pending
- Do not allow new association requests if there is a pending request from a peer / peer already associated
- Only create new requests if the last request from the peer has been denied or the peer does not submit any request yet

Reply:
Looks good to me apart from a nitpick change. Needs a second pair of eyes from @shubham3121 too.